File: C++/Minimum cost to fill given weight in a bag.cpp
class Solution {
  public:
    int minimumCost(int n, int w, vector<int> &cost) {
        vector<int> dp(w + 1, INT_MAX);
        dp[0] = 0;
        for(int i = 1; i <= w; i++) {
            for(int j = 0; j < n; j++) {
                if(cost[j] != -1 && (i - j - 1) >= 0) {
                    dp[i] = min(dp[i], cost[j] + dp[i - j - 1]);
                }
            }
        }
        return dp[w];
    }
};

File: JAVA/Minimum cost to fill given weight in a bag.java

class Solution {
    public static int minimumCost(int n, int w, int[] cost) {
         HashMap<String,Integer> hs=new HashMap<>();
        int a = solve(n,1,w,cost,hs);
        return (a==80000)?-1:a;
    }
    public static int solve(int n,int i,int w,int arr[],HashMap<String,Integer> hs){
        if(w==0)
            return 0;
        if(i-1==arr.length)
            return 80000;
        String key=i+"nik"+w;
        if(hs.containsKey(key))
            return hs.get(key);
        int a=80000,b=80000;
        if(w-i>=0 && arr[i-1]!=-1)
            a=arr[i-1]+solve(n,i,w-i,arr,hs);
        b=solve(n,i+1,w,arr,hs);
        int min = Math.min(a,b);
        hs.put(key,min);
        return min;
    }
}
